agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/TelemetryChannel.java

```diff
@@ -302,47 +309,52 @@ private void writeToDiskOnFailure(List<ByteBuffer> byteBuffers, String instrumen
     }
   }
 
-  private void parseResponseCode(
-      int statusCode, String instrumentationKey, List<ByteBuffer> byteBuffers, boolean persisted) {
+  private void parseResponseCode(int statusCode, String body, String instrumentationKey) {
     switch (statusCode) {
+      case 200: // SUCCESS
+        operationLogger.recordSuccess();
+        break;
+      case 206: // PARTIAL CONTENT, Breeze-specific: PARTIAL SUCCESS
+        operationLogger.recordFailure(getErrorMessageFromPartialSuccessResponse(body));
+        break;
       case 401: // UNAUTHORIZED
       case 403: // FORBIDDEN
-        logger.warn(
-            "Failed to send telemetry with status code:{}, please check your credentials",
-            statusCode);
-        // no need to write to disk again when failing to send raw bytes from the persisted file
-        if (!persisted) {
-          writeToDiskOnFailure(byteBuffers, instrumentationKey);
-        }
-        break;
       case 408: // REQUEST TIMEOUT
       case 500: // INTERNAL SERVER ERROR
       case 503: // SERVICE UNAVAILABLE
       case 429: // TOO MANY REQUESTS
+        // TODO (heya) should we write to disk on any of these response codes?
+        operationLogger.recordFailure("received response code " + statusCode);
+        break;
       case 439: // Breeze-specific: THROTTLED OVER EXTENDED TIME
+        operationLogger.recordFailure("received response code 439 (throttled over extended time)");
         // TODO handle throttling
         // TODO (heya) track throttling count via Statsbeat
         // instrumentationKey is null when sending persisted file's raw bytes.
         if (!isStatsbeat) {
           statsbeatModule.getNetworkStatsbeat().incrementThrottlingCount(instrumentationKey);
         }
         break;
-      case 200: // SUCCESS
-        operationLogger.recordSuccess();
-        break;
-      case 206: // PARTIAL CONTENT, Breeze-specific: PARTIAL SUCCESS
-        // TODO handle partial success
-        break;
-      case 0: // client-side exception
-        // TODO exponential backoff and retry to a limit
-        // TODO (heya) track failure count via Statsbeat
-        // instrumentationKey is null when sending persisted file's raw bytes.
-        if (!isStatsbeat) {
-          statsbeatModule.getNetworkStatsbeat().incrementRetryCount(instrumentationKey);
-        }
-        break;
       default:
-        // ok
+        operationLogger.recordFailure("received response code: " + statusCode);
+    }
+  }
+
+  private static String getErrorMessageFromPartialSuccessResponse(String body) {
```

Comment:
i wonder how do you generate 206 error code? i haven't run into it..

Reply:
see #2064

I ran into it doing some testing yesterday:

```
Span span = tracer.spanBuilder("test").startSpan();
span.recordException(new Throwable());
span.end();
```